spring-core/src/main/java/org/springframework/core/codec/CharSequenceEncoder.java

```diff
@@ -75,12 +79,32 @@ public boolean canEncode(ResolvableType elementType, @Nullable MimeType mimeType
 					return Hints.getLogPrefix(hints) + "Writing " + formatted;
 				});
 			}
-			CharBuffer charBuffer = CharBuffer.wrap(charSequence);
-			ByteBuffer byteBuffer = charset.encode(charBuffer);
-			return bufferFactory.wrap(byteBuffer);
+			boolean release = true;
+			int capacity = calculateCapacity(charSequence, charset);
+			DataBuffer dataBuffer = bufferFactory.allocateBuffer(capacity);
+			try {
+				dataBuffer.write(charSequence, charset);
+				release = false;
+			}
+			catch (IllegalStateException ex) {
+				throw new EncodingException("String encoding error: " + ex.getMessage(), ex);
+			}
+			finally {
+				if (release) {
+					DataBufferUtils.release(dataBuffer);
+				}
+			}
+			return dataBuffer;
 		});
 	}
 
+	private int calculateCapacity(CharSequence sequence, Charset charset) {
```

Comment:
I think I would like to see a separate unit test for this method, but I can add that test myself as well.